src/libos/src/vm/vm_area.rs

```diff
@@ -202,6 +485,292 @@ impl VMArea {
+            return_errno!(EBADF, "file is not readable");
+        }
+
+        let buf = unsafe { target_range.as_slice_mut() };
+        let file_size = file.metadata().unwrap().size;
+
+        let len = file
+            .read_at(file_offset, buf)
+            .map_err(|_| errno!(EACCES, "failed to init memory from file"))?;
+        trace!("file offset = {:?}, read len = {:?}", file_offset, len);
+        trace!("file total size = {:?}", file_size);
+
+        if !new_perm.is_default() {
+            self.modify_protection_force(Some(target_range), new_perm);
+        }
+
+        Ok(())
+    }
+
+    // Inintialize the VMA memory if the VMA is partially committed
+    fn init_committed_memory(&mut self, initializer: &VMInitializer) -> Result<()> {
+        debug_assert!(self.is_partially_committed());
+        let committed = true;
+        let is_handle_pf = false;
+        for range in self.pages().get_ranges(committed) {
+            trace!("init committed memory: {:?}", range);
+            self.init_committed_memory_internal(&range, Some(initializer))?;
+        }
+
+        Ok(())
+    }
+
+    pub fn flush_committed_memory(&self) -> Result<()> {
+        debug_assert!(self.is_partially_committed());
+        trace!("flush committed memory");
+
+        let (need_flush, file, file_offset) = match self.writeback_file() {
+            None => (false, None, None),
+            Some((file_handle, offset)) => {
+                if !file_handle.access_mode().unwrap().writable() {
+                    (false, None, None)
+                } else {
+                    (true, Some(file_handle), Some(offset))
+                }
+            }
+        };
+
+        let vm_range_start = self.range.start();
+        let committed = true;
+        for range in self.pages().get_ranges(committed) {
+            info!("flush committed range: {:?}", range);
+            let buf = unsafe { range.as_slice_mut() };
+            if !self.perms().is_default() {
+                self.modify_protection_force(Some(&range), VMPerms::default());
+            }
+
+            if need_flush {
+                let file_offset = file_offset.unwrap() + (range.start() - vm_range_start);
+                file.unwrap().write_at(file_offset, buf);
+            }
+
+            // reset zeros
+            trace!("reset zeros for range: {:?}", range);
+            unsafe {
+                buf.iter_mut().for_each(|b| *b = 0);
+            }
+        }
+        Ok(())
+    }
+
+    pub fn commit_once_for_page_fault(&mut self, pf_addr: usize) -> Result<usize> {
+        debug_assert!(!self.is_fully_committed());
+        let mut early_return = false;
+        let mut total_commit_size = 0;
+        let vma_range_start = self.range.start();
+        let permission = self.perms();
+        let committed = false;
+        let mut uncommitted_ranges = self.pages().get_ranges(committed);
+
+        for range in uncommitted_ranges.iter_mut() {
```

Comment:
It seems that the `skip_while` will always skip the range which does not contain the `pf_addr`. However, this is incorrect because only the first iteration should do this check. I did some changes accordingly to make it more concise. Please take a look. Thank you for your suggestions.